File: SRM 755/OneHandRoadPainting.cpp
#include <cstdio>
#include <cmath>
#include <cstring>
#include <ctime>
#include <iostream>
#include <algorithm>
#include <set>
#include <vector>
#include <sstream>
#include <typeinfo>
#include <fstream>

using namespace std;

class OneHandRoadPainting {

  public:
    long long solve(vector<pair<int, int> > vzip, int paintPerBrush)
    {
      int n = vzip.size();
      long long ans = 0;
      for (int i = 0; i < n; i++)
      {
        div_t res = div(vzip[i].first - vzip[i].second, paintPerBrush);
        int numseg = res.quot, restseg = res.rem;
        if (numseg == 0 && restseg == 0) ans += 0; // This case is not possilbe
        else if (numseg != 0 && restseg == 0)
            ans = ans + (long long) 2 * sumAP(vzip.[i].first, numseg, -paintPerBrush);
        else if (numseg == 0 && restseg != 0)
        {
            


      }
      return 0;

    }

    long long fastest(vector<int> dStart, vector<int> dEnd, int paintPerBrush) {
      vector<pair<int, int> > vzip;
      int n = dStart.size();
      for (int i = 0; i <= n; i++) vzip.push_back(make_pair(dEnd[i], dStart[i]));
      reverse(vzip.begin(), vzip.end());
      return solve(vzip, paintPerBrush);
    }
};

// CUT begin
ifstream data("OneHandRoadPainting.sample");

string next_line() {
    string s;
    getline(data, s);
    return s;
}

template <typename T> void from_stream(T &t) {
    stringstream ss(next_line());
    ss >> t;
}

void from_stream(string &s) {
    s = next_line();
}

template <typename T> void from_stream(vector<T> &ts) {
    int len;
    from_stream(len);
    ts.clear();
    for (int i = 0; i < len; ++i) {
        T t;
        from_stream(t);
        ts.push_back(t);
    }
}

template <typename T>
string to_string(T t) {
    stringstream s;
    s << t;
    return s.str();
}

string to_string(string t) {
    return "\"" + t + "\"";
}

bool do_test(vector<int> dStart, vector<int> dEnd, int paintPerBrush, long long __expected) {
    time_t startClock = clock();
    OneHandRoadPainting *instance = new OneHandRoadPainting();
    long long __result = instance->fastest(dStart, dEnd, paintPerBrush);
    double elapsed = (double)(clock() - startClock) / CLOCKS_PER_SEC;
    delete instance;

    if (__result == __expected) {
        cout << "PASSED!" << " (" << elapsed << " seconds)" << endl;
        return true;
    }
    else {
        cout << "FAILED!" << " (" << elapsed << " seconds)" << endl;
        cout << "           Expected: " << to_string(__expected) << endl;
        cout << "           Received: " << to_string(__result) << endl;
        return false;
    }
}

int run_test(bool mainProcess, const set<int> &case_set, const string command) {
    int cases = 0, passed = 0;
    while (true) {
        if (next_line().find("--") != 0)
            break;
        vector<int> dStart;
        from_stream(dStart);
        vector<int> dEnd;
        from_stream(dEnd);
        int paintPerBrush;
        from_stream(paintPerBrush);
        next_line();
        long long __answer;
        from_stream(__answer);

        cases++;
        if (case_set.size() > 0 && case_set.find(cases - 1) == case_set.end())
            continue;

        cout << "  Testcase #" << cases - 1 << " ... ";
        if ( do_test(dStart, dEnd, paintPerBrush, __answer)) {
            passed++;
        }
    }
    if (mainProcess) {
        cout << endl << "Passed : " << passed << "/" << cases << " cases" << endl;
        int T = time(NULL) - 1555841312;
        double PT = T / 60.0, TT = 75.0;
        cout << "Time   : " << T / 60 << " minutes " << T % 60 << " secs" << endl;
        cout << "Score  : " << 500 * (0.3 + (0.7 * TT * TT) / (10.0 * PT * PT + TT * TT)) << " points" << endl;
    }
    return 0;
}

int main(int argc, char *argv[]) {
    cout.setf(ios::fixed, ios::floatfield);
    cout.precision(2);
    set<int> cases;
    bool mainProcess = true;
    for (int i = 1; i < argc; ++i) {
        if ( string(argv[i]) == "-") {
            mainProcess = false;
        } else {
            cases.insert(atoi(argv[i]));
        }
    }
    if (mainProcess) {
        cout << "OneHandRoadPainting (500 Points)" << endl << endl;
    }
    return run_test(mainProcess, cases, argv[0]);
}
// CUT end

File: SRM 755/OneHandRoadPainting.py
# -*- coding: utf-8 -*-
import math,string,itertools,fractions,heapq,collections,re,array,bisect

#Python is way more cleaner than C++
class OneHandRoadPainting:

    def sumAP(self, a, n, d):
            return  (n * (2 * a + (n - 1) * d)) / 2

    def sliceSeg(self, izip, paintPerBrush):
        if (izip == []): return izip
        (endseg, begseg), rest = izip[0], izip[1::]
        if (paintPerBrush >= endseg - begseg):
            return self.sliceSeg(izip[1::], paintPerBrush - (endseg - begseg))
        else:
            return [tuple ((endseg - paintPerBrush, begseg))] + izip[1::]


    def solve(self, izip, paintPerBrush):
        if (izip == []):
            return 0
        else:
            (endseg, begseg), rest = izip[0], izip[1::]
            (numseg, restseg) = divmod(endseg - begseg, paintPerBrush)
            if numseg == 0 and restseg == 0 : 
                raise Exception("This should not have had happened")
            elif numseg != 0 and restseg == 0 :
                return (2 * self.sumAP(endseg, numseg, -paintPerBrush) 
                        + self.solve(izip[1::], paintPerBrush))
            elif numseg == 0 and restseg != 0 :
                #slice the part which can be painted
                rest = self.sliceSeg(izip, paintPerBrush)
                return (2 * endseg + self.solve(rest, paintPerBrush))
            else:
                #get rid of painted segment and simulat the rest
                tzip = [tuple ((endseg - numseg * paintPerBrush, begseg))] + izip[1::]
                rest = self.sliceSeg(tzip, paintPerBrush)
                return (2 * self.sumAP(endseg, numseg, -paintPerBrush) + 
                        2 * (endseg - numseg * paintPerBrush) + self.solve(rest, paintPerBrush))




    def fastest(self, dStart, dEnd, paintPerBrush):
        izip = zip(dEnd, dStart)[::-1] #zip the list, and reverse it
        return self.solve(izip, paintPerBrush)


# CUT begin
# TEST CODE FOR PYTHON {{{
import sys, time, math

def tc_equal(expected, received):
    try:
        _t = type(expected)
        received = _t(received)
        if _t == list or _t == tuple:
            if len(expected) != len(received): return False
            return all(tc_equal(e, r) for (e, r) in zip(expected, received))
        elif _t == float:
            eps = 1e-9
            d = abs(received - expected)
            return not math.isnan(received) and not math.isnan(expected) and d <= eps * max(1.0, abs(expected))
        else:
            return expected == received
    except:
        return False

def pretty_str(x):
    if type(x) == str:
        return '"%s"' % x
    elif type(x) == tuple:
        return '(%s)' % (','.join( (pretty_str(y) for y in x) ) )
    else:
        return str(x)

def do_test(dStart, dEnd, paintPerBrush, __expected):
    startTime = time.time()
    instance = OneHandRoadPainting()
    exception = None
    try:
        __result = instance.fastest(dStart, dEnd, paintPerBrush);
    except:
        import traceback
        exception = traceback.format_exc()
    elapsed = time.time() - startTime   # in sec

    if exception is not None:
        sys.stdout.write("RUNTIME ERROR: \n")
        sys.stdout.write(exception + "\n")
        return 0

    if tc_equal(__expected, __result):
        sys.stdout.write("PASSED! " + ("(%.3f seconds)" % elapsed) + "\n")
        return 1
    else:
        sys.stdout.write("FAILED! " + ("(%.3f seconds)" % elapsed) + "\n")
        sys.stdout.write("           Expected: " + pretty_str(__expected) + "\n")
        sys.stdout.write("           Received: " + pretty_str(__result) + "\n")
        return 0

def run_tests():
    sys.stdout.write("OneHandRoadPainting (500 Points)\n\n")

    passed = cases = 0
    case_set = set()
    for arg in sys.argv[1:]:
        case_set.add(int(arg))

    with open("OneHandRoadPainting.sample", "r") as f:
        while True:
            label = f.readline()
            if not label.startswith("--"): break

            dStart = []
            for i in range(0, int(f.readline())):
                dStart.append(int(f.readline().rstrip()))
            dStart = tuple(dStart)
            dEnd = []
            for i in range(0, int(f.readline())):
                dEnd.append(int(f.readline().rstrip()))
            dEnd = tuple(dEnd)
            paintPerBrush = int(f.readline().rstrip())
            f.readline()
            __answer = int(f.readline().rstrip())

            cases += 1
            if len(case_set) > 0 and (cases - 1) in case_set: continue
            sys.stdout.write("  Testcase #%d ... " % (cases - 1))
            passed += do_test(dStart, dEnd, paintPerBrush, __answer)

    sys.stdout.write("\nPassed : %d / %d cases\n" % (passed, cases))

    T = time.time() - 1555841226
    PT, TT = (T / 60.0, 75.0)
    points = 500 * (0.3 + (0.7 * TT * TT) / (10.0 * PT * PT + TT * TT))
    sys.stdout.write("Time   : %d minutes %d secs\n" % (int(T/60), T%60))
    sys.stdout.write("Score  : %.2f points\n" % points)

if __name__ == '__main__':
    run_tests()

# }}}
# CUT end
